adapters/sspbc/sspbc.go

```diff
@@ -0,0 +1,407 @@
+				var adCreationError error
+
+				// Prepare ads (using different methods for banner, native, video)
+
+				// BANNER
+				bid.AdM, adCreationError = a.createBannerAd(bid, BidDataExt, internalRequest, seatBid.Seat)
+
+				if adCreationError != nil {
+					glog.Errorf("SSPBC: cannot format creative")
+					errors = append(errors, err)
+				} else {
+					// append bid to responses
+					bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
+						Bid:     &bid,
+						BidType: BidType,
+					})
+				}
+			}
+		}
+	}
+
+	return bidResponse, errors
+}
+
+func (a *SspbcAdapter) createBannerAd(bid openrtb2.Bid, ext SsbcResponseExt, request *openrtb2.BidRequest, seat string) (string, error) {
+	var mcad McAd
+
+	if strings.Contains(bid.AdM, "<!--preformatted-->") {
+		// Banner ad is already formatted
+		return bid.AdM, nil
+	}
+
+	// create McAd payload
+	mcad.Id = request.ID
+	mcad.Seat = seat
+	mcad.SeatBid = make([]openrtb2.SeatBid, 1)
+	mcad.SeatBid[0].Bid = make([]openrtb2.Bid, 1)
+	mcad.SeatBid[0].Bid[0] = bid
+	mcMarshalled, err := json.Marshal(mcad)
+	if err != nil {
+		glog.Errorf("SSPBC: Cannot Marshal mcad!")
+		return bid.AdM, err
+	}
+
+	mcEncoded := base64.URLEncoding.EncodeToString(mcMarshalled)
+
+	bannerData := &TemplatePayload{
+		SiteId:  ext.SiteId,
+		SlotId:  ext.SlotId,
+		AdLabel: ext.AdLabel,
+		PubId:   ext.PublisherId,
+		Page:    request.Site.Page,
+		Referer: request.Site.Ref,
+		McAd:    mcEncoded,
+	}
+
+	/*
+		Prepare banner html, using template file
+
+		Note: Prebidserver bidders have access only to gdpr data in user ext, which is not what we need (as mcad uses prebid.js gdpr format)
+		Therefore, we are not creating window.gdpr. This will force banner creative to execute it's own call to TCF2
+	*/
+
+	var filledTemplate bytes.Buffer
+	if err := a.bannerTemplate.Execute(&filledTemplate, bannerData); err != nil {
+		glog.Errorf("SSPBC: Cannot execute banner template")
+		return bid.AdM, err
+	}
+
+	/*
+		byteTemplate := []byte(filledTemplate.String())
+		fmt.Println(base64.StdEncoding.EncodeToString(byteTemplate)) */
+
+	return filledTemplate.String(), nil
+}
+
+func getImpSize(Imp openrtb2.Imp) string {
+	if Imp.Video != nil {
+		return fmt.Sprintf("%dx%d", Imp.Video.W, Imp.Video.H)
+	}
+
+	if Imp.Banner != nil {
+		sb := strings.Builder{}
+		areaMax := int64(0)
+		sb.WriteString("1x1")
+		for _, sizeI := range Imp.Banner.Format {
+			areaI := sizeI.W * sizeI.H
+			if areaI > areaMax {
+				areaMax = areaI
+				sb.Reset()
+				sb.WriteString(fmt.Sprintf("%dx%d", sizeI.W, sizeI.H))
+			}
+		}
+		return sb.String()
+	}
+
+	// default fallback
+	return "1x1"
+}
+
+func formatSsbcRequest(a *SspbcAdapter, request *openrtb2.BidRequest) (*openrtb2.BidRequest, error) {
+	var err error
+	var siteId string
+	var isTest int
+
+	// check if adSlots and adSizes maps are initialized
+	if a.adSlots == nil {
+		a.adSlots = make(map[string]AdSlotData)
+	}
+	if a.adSizes == nil {
+		a.adSizes = make(map[string]int)
+	}
```

Comment:
Why are adSlots and adSizes global to your adapter?

Reply:
adSlots and adSizes are used between auctions to determine siteid & slotid, if those are not present in bid params.
Basically we are mapping those id's to either adSlot name, or AdSize x AdSizeUsedCount

Reply:
There is only ever once instance of the adapter in memory. How does the adSlot and adSizes persistent map work across different publishers?

Reply:
Please take a look at below convo. This is the equivalent of additional slot info that we're doing in js adapter, but - as @hhhjort pointed out - it will not work if there's more than one instance of PBS running. This is now simplified (and functionality that it provided will have to be provided as first party data)